### PR TITLE
Add endpoint to get rates by state and date

### PR DESCRIPTION
**Pull request by coconut750750**
- Refactor some code to get response rates

- Add endpoint to get response rates by state and date

- Add endpoint to get response rates by year

- Add optional `tract_id` parameter to specifically get rate by tract_id

### CODE REVIEW

File: backend/api/views/response_rates.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        response_rates[resp.tract_id] = end_rate
+
+    return [{"tract_id": tid, "rate": rate} for tid, rate in response_rates.items()]
+
+
+def get_response_rates_by_state(state, date):
+    response_rates = {}
+
+    regex = re.compile("^{}.*".format(state))
+    logger.info(regex)
```
Comment (chyku) [inline backend/api/views/response_rates.py]: leftover log?

File: backend/api/views/main.py [hunk 2]
```diff
@@ -74,6 +52,26 @@ def create_person():
     )
 
 
+# function that is called when you visit /census_response
```
Comment (chyku) [inline backend/api/views/main.py]: outdated comment, and also what form is the state going to be passed into this, as a 2-digit ID? should also be commented

File: backend/api/views/main.py [hunk 3]
```diff
@@ -74,6 +52,26 @@ def create_person():
...
+            response_rates = get_response_rates_by_state(request.args["state"], date)
+        else:
+            response_rates = get_response_rates_by_date(date)
+
+    elif "year" in request.args:
+        response_rates = get_response_rates_by_year(request.args["year"])
+    else:
+        return create_response(status=442, message="Missing request parameters")
+
+    return create_response(data={"response_rates": response_rates})
```
Comment (chyku) [inline backend/api/views/main.py]: We should also document the specific combinations of parameters that we're supporting

File: backend/api/views/response_rates.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+from api.core import logger
+from api.models import db, CensusResponse
+import re
+
+
+def get_response_rates_by_date(date):
+    response_rates = {}
+
+    responses = CensusResponse.objects()
+    year = date[-4:]
```
Comment (chyku) [inline backend/api/views/response_rates.py]: should document the format the date should be in

File: backend/api/views/response_rates.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        end_rate = rates[max(rates, key=str)]
+        response_rates[resp.tract_id] = end_rate
+
+    return [{"tract_id": tid, "rate": rate} for tid, rate in response_rates.items()]
+
+
+def get_response_rates_by_state(state, date):
+    response_rates = {}
+
+    regex = re.compile("^{}.*".format(state))
```
Comment (chyku) [inline backend/api/views/response_rates.py]: what is this searching for and what does this do? plz comment

**Review by chyku — CHANGES_REQUESTED** [review]
lotsa comments plz

File: backend/api/views/main.py [hunk 6]
```diff
@@ -74,6 +53,34 @@ def create_person():
...
 
 
+'''
+function that is called when you visit /response_rates
+Parameters
+    date: date string with the format MMDDYYYY
+    year: year string with format YY
+    optional tract_id: 11-digit tract id string
+    optional state: two digit id string
+Either date or year is required.
```
Comment (chyku) [inline backend/api/views/main.py:62]: 😍 bless this documentation

File: backend/api/views/main.py [hunk 7]
```diff
@@ -74,6 +53,34 @@ def create_person():
...
+    date = request.args.get("date", None)
+    year = request.args.get("year", None)
+    state = request.args.get("state", None)
+
+    if date:
+        response_rates = get_response_rates_by_date(date, tract_id, state)
+    elif year:
+        response_rates = get_response_rates_by_year(year, tract_id, state)
+    else:
+        return create_response(status=442, message="Missing request parameters")
```
Comment (chyku) [inline backend/api/views/main.py]: is this a valid response code or do you mean 422

File: backend/api/views/response_rates.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+def get_year(date):
+    return str(time.strptime(date, "%m%d%Y").tm_year)
+
+'''
+returns a collection of CensusResponses
+filtered by tract_id if tract_id parameter is not None
+'''
+def get_census_responses(tract_id, state):
+    if state:
+        regex = re.compile("^{}.*".format(state))
```
Comment (chyku) [inline backend/api/views/response_rates.py]: pls comment what this does for readability

**Review by chyku — CHANGES_REQUESTED** [review]
*chef hand kiss* beautiful comments + a couple of qs

File: backend/api/views/main.py (hunk 7 above)
Comment (coconut750750) [reply backend/api/views/main.py]: lol i meant 422

**Review by caiyingshen — APPROVED** [review]
beep beep

File: backend/api/views/main.py (hunk 6 above)
Comment (coconut750750) [reply backend/api/views/main.py:62]: 😍

File: backend/api/views/response_rates.py (hunk 8 above)
Comment (coconut750750) [reply backend/api/views/response_rates.py]: yup my bad added it back

**Review by chyku — APPROVED** [review]
😩😩👏👏👏

**Pull request merged**